Fix double instances of legacy SEPA gateway being loaded

james-allan opened the pull request:
Fixes #3247

## Changes proposed in this Pull Request:

With the legacy checkout experience enabled, if you go to add a SEPA token via the **My Account → Payments methods**, you will see 2 instances of the same checkbox.

<p align="center">
<img width="600" alt="Screenshot 2024-07-05 at 1 28 04 PM" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/c8a0b9f8-2603-4e57-9cd9-43abf3ef49ab">
</p>

This PR fixes that.

It turns out that there were 2 instances of the `WC_Gateway_Stripe_Sepa` class being instantiated and that caused attached hooks to double up. The two instances were from WC creating an instance of the class when we return the class here, and then the second instance was in our code here. This PR fixes that by making sure we use the instance of the gateway already in use, rather than creating a new instance.

## Testing instructions

1. Enable the **Legacy checkout experience** in Stripe plugin settings.
2. Set your store to EUR.
3. Enable SEPA payment methods.
4. Go to you **My Account → Payments methods**
5. **Add payment method**.
6. Select SEPA and notice that there are 2 checkbox (see screenshot above).
7. Checkout this branch and confirm that there's only 1 checkbox. 

<p align="center">
<img width="600" alt="Screenshot 2024-07-05 at 1 39 53 PM" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/c6a7e310-0b35-405f-bfbf-7a72c054a3f3">
</p>


---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

james-allan committed:
Use the already existing instance of legacy gateways rather than creating new ones

james-allan committed:
Remove array flip and use in_array and search

james-allan committed:
Use strict search

james-allan committed:
Add changelog entry

james-allan committed:
Restore original variable name

wjrosa reviewed (changes requested):
Hi James! I am having a weird issue with this change. That's how my payment methods page looks like when I use this branch (Stripe is gone):
![Screenshot 2024-07-05 at 18 58 56](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/10187816/52455bdc-dda5-4f6c-9854-459d0047965c)

I double-checked using the `develop` branch and did a manual revert of your change on your branch. In both cases Stripe shows up again 🤔

james-allan committed:
Don't call payment_gateways() if the gateways haven't been initialized yet

james-allan committed:
Rework how the legacy Stripe gateway is initalized to not call get_legacy_enabled_payment_method_ids()

james-allan commented:
> That's how my payment methods page looks like when I use this branch (Stripe is gone):

Nice spotting. I did not consider a change like this could introduce a bug like that 😕.
 
I spent a bit of time digging into what's causing that and it turns out that it's caused by calling `WC()->payment_gateways()->payment_gateways()` too early. 

When you have the legacy experience enabled and the Stripe plugin is being initialised, we were calling `get_legacy_enabled_payment_method_ids()` which would eventually lead to `get_legacy_payment_methods()`. In this PR I changed that function to call `WC()->payment_gateways()->payment_gateways()`. 

That means that we are now essentially calling `WC()->payment_gateways()->payment_gateways()` on the Stripe plugin init.

Doing that so early caused WC to initialise the gateways and cache them in memory before other gateways have had a chance to register their gateways via the `woocommerce_payment_gateways` filter. I've included the full stack trace below.

<details><summary>Full stacktrace</summary>
<pre>
#0 /wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-helper.php(527): WC_Stripe_Helper::get_legacy_payment_methods()
#1 /wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-gateway-stripe.php(116): WC_Stripe_Helper::get_legacy_enabled_payment_method_ids()
#2 /wp-content/plugins/woocommerce-gateway-stripe/woocommerce-gateway-stripe.php(705): WC_Gateway_Stripe->__construct()
#3 /wp-content/plugins/woocommerce-gateway-stripe/woocommerce-gateway-stripe.php(262): WC_Stripe->get_main_stripe_gateway()
#4 /wp-content/plugins/woocommerce-gateway-stripe/woocommerce-gateway-stripe.php(164): WC_Stripe->init()
#5 /wp-content/plugins/woocommerce-gateway-stripe/woocommerce-gateway-stripe.php(101): WC_Stripe->__construct()
#6 /wp-content/plugins/woocommerce-gateway-stripe/woocommerce-gateway-stripe.php(751): WC_Stripe::get_instance()
#7 /wp-content/plugins/woocommerce-gateway-stripe/woocommerce-gateway-stripe.php(773): woocommerce_gateway_stripe()
#8 /wp-includes/class-wp-hook.php(324): woocommerce_gateway_stripe_init('')
#9 /wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters(NULL, Array)
#10 /wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#11 /wp-settings.php(550): do_action('plugins_loaded')
#15 {main}
</pre>
</details> 

I've fixed this (in [867b54f](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3248/commits/867b54ff14b07cc9769d3f257ed23809229cc372)) by moving the logic which changes the gateway's title on the settings screen - this is the logic that was calling `get_legacy_enabled_payment_method_ids()` and leading to us initialising the gateways. I've move that logic into the `get_title()` so that logic only runs when it's strictly needed rather than on gateway init.

james-allan committed:
Merge branch 'develop' into fix/double-instances-of-legacy-sepa-gateway

james-allan committed:
Tweak comments

wjrosa reviewed (approved):
That fix was quick! I tested it again, and is all good now. Also, the main fix worked as well:
![Screenshot 2024-07-08 at 11 18 57](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/10187816/26fdb909-3979-4b2f-b2ec-d546e50b83fe)

james-allan committed:
Merge branch 'develop' into fix/double-instances-of-legacy-sepa-gateway

james-allan committed:
Remove extra space added by error
